csharp/src/Ice/SlicBinaryConnection.cs

```diff
@@ -107,164 +161,233 @@ public async ValueTask CloseAsync(Exception exception, CancellationToken cancel)
+                writer!(ostr);
             }
+            ostr.EndFixedLengthSize(sizePos);
+            data[^1] = data[^1].Slice(0, ostr.Tail.Offset); // TODO: Shouldn't this be the job of ostr.Finish()?
 
-            if (size > _frameSizeMax)
+            if (Endpoint.Communicator.TraceLevels.Protocol >= 1)
             {
-                throw new InvalidDataException($"frame with {size} bytes exceeds Ice.IncomingFrameSizeMax value");
+                TraceFrame("sending ", type, data.GetByteCount());
             }
 
-            // Read the remainder of the frame if needed
-            if (size > readBuffer.Count)
-            {
-                if (size > readBuffer.Array!.Length)
-                {
-                    // Allocate a new array and copy the header over
-                    var buffer = new ArraySegment<byte>(new byte[size], 0, size);
-                    readBuffer.AsSpan().CopyTo(buffer.AsSpan(0, Ice2Definitions.HeaderSize));
-                    readBuffer = buffer;
-                }
-                else
-                {
-                    readBuffer = new ArraySegment<byte>(readBuffer.Array!, 0, size);
-                }
-                Debug.Assert(size == readBuffer.Count);
-
-                await ReceiveAsync(readBuffer.Slice(Ice2Definitions.HeaderSize)).ConfigureAwait(false);
-            }
-            return readBuffer;
+            await SendSlicFrameAsync(data, cancel).ConfigureAwait(false);
         }
 
-        private async ValueTask PerformSendFrameAsync(long streamId, object frame)
+        private async ValueTask<(FrameType, ArraySegment<byte>)> ReceiveFrameAsync(CancellationToken cancel)
         {
-            List<ArraySegment<byte>> writeBuffer;
+            ArraySegment<byte> buffer = await _transceiver.ReceiveAsync(2, cancel).ConfigureAwait(false);
+            var frameType = (FrameType)buffer[0];
+            byte firstSizeByte = buffer[1];
+            int sizeLength = firstSizeByte.ReadSizeLength20();
+            buffer = await _transceiver.ReceiveAsync(sizeLength - 1, cancel).ConfigureAwait(false);
+            _receivedCallback(sizeLength + 1);
 
-            // TODO: add abstract OutgoingFrame class with an abstract GetRequestData(streamId) method?
-            if (frame is OutgoingRequestFrame requestFrame)
+            int frameSize = ComputeSize20(firstSizeByte, buffer.AsReadOnlySpan());
+
+            ArraySegment<byte> frame = new byte[frameSize];
+            int offset = 0;
+            while (offset != frameSize)
             {
-                writeBuffer = Ice2Definitions.GetRequestData(requestFrame, streamId);
-                ProtocolTrace.TraceFrame(Endpoint.Communicator, writeBuffer[0], requestFrame);
+                int received = await _transceiver.ReceiveAsync(frame.Slice(offset), cancel).ConfigureAwait(false);
+                offset += received;
+                _receivedCallback!(received);
```

Comment:
Why this `!`?

Reply:
I forgot to fix the data member to be optional, fixed now.